src/Bicep.Core/Semantics/YamlObjectParser.cs

```diff
@@ -26,7 +38,45 @@ public class YamlObjectParser : ObjectParser
         }
 
         override protected Diagnostic GetExtractTokenErrorType(IPositionable positionable)
-            => DiagnosticBuilder.ForPosition(positionable).UnparsableYamlType();
+        {
+            return isMultiDocumentYaml 
+                ? DiagnosticBuilder.ForPosition(positionable).MultiDocumentYamlNotSupported()
+                : DiagnosticBuilder.ForPosition(positionable).UnparsableYamlType();
+        }
 
+        // Helper method to detect multi-document YAML
+        private bool IsMultiDocumentYaml(string content)
+        {
+            using var reader = new StringReader(content);
+            string? line;
+            bool foundFirstDocument = false;
+            
+            while ((line = reader.ReadLine()) != null)
+            {
+                // Skip empty lines and comments
+                if (string.IsNullOrWhiteSpace(line) || line.TrimStart().StartsWith('#'))
+                {
+                    continue;
+                }
+                    
+                // Check for document separator
+                if (line.Trim() == "---")
```

Comment:
How fragile of a check is this? I think there are instances where `---` alone on a line is not a document separator (though admittedly this may be a bug in [YAMLine](https://yamline.com/json/) rather than correct according to the spec).

```yaml
comments: >
  ---
  ---
array: [
  foo,
  ---
]
```

Ideally, I would want the YAML parser we're using to read the file to report back whether it contains multiple documents rather than rescanning the file and doing our own check.